src/cmap/conn/mod.rs

```diff
@@ -340,6 +396,40 @@ impl Drop for Connection {
     }
 }
 
+/// A handle to a pinned connection - the connection itself can be retrieved or returned to the
+/// normal pool via this handle.
+#[derive(Debug)]
+pub(crate) struct PinnedConnectionHandle {
+    receiver: Arc<Mutex<mpsc::Receiver<Connection>>>,
+}
+
+impl PinnedConnectionHandle {
+    /// Make a new `PinnedConnectionHandle` that refers to the same connection as this one.
+    /// Use with care and only when "lending" a handle in a way that can't be expressed as a
+    /// normal borrow.
+    pub(crate) fn replicate(&self) -> Self {
+        Self {
+            receiver: self.receiver.clone(),
+        }
+    }
+
+    /// Retrieve the pinned connection, blocking until it's available for use.  Will fail if the
+    /// connection has been unpinned.
+    pub(crate) async fn take_connection(&self) -> Result<Connection> {
+        let mut receiver = self.receiver.lock().await;
+        receiver
+            .recv()
+            .await
+            .ok_or_else(|| Error::internal("cannot take connection after unpin"))
+    }
+
+    /// Return the pinned connection to the normal connection pool.
+    pub(crate) async fn unpin_connection(&self) {
+        let mut receiver = self.receiver.lock().await;
+        receiver.close();
```

Comment:
Per the [tokio docs](https://docs.rs/tokio/1.11.0/tokio/sync/mpsc/struct.Receiver.html), this won't drop the connection if its in the channel, so we need to drain it to ensure the connection gets dropped and returned to the pool.

Reply:
Done.